library/std/src/path.rs

```diff
@@ -1973,7 +1973,27 @@ impl Path {
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn new<S: AsRef<OsStr> + ?Sized>(s: &S) -> &Path {
-        unsafe { &*(s.as_ref() as *const OsStr as *const Path) }
+        Self::from_os_str(s.as_ref())
+    }
+
+    /// Creates a new [`Path`] from an [`OsStr`].
+    ///
+    /// This method supports const expressions. However, if you don't need const,
+    /// you should probably use the [`Path::new`] method instead.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(const_path)]
+    /// use std::ffi::OsStr;
+    /// use std::path::Path;
+    ///
+    /// const PATH: &Path = Path::from_os_str(OsStr::from_str("/foo/bar"));
+    /// ```
+    #[inline]
+    #[unstable(feature = "const_path", reason = "TBD", issue = "none")]
+    pub const fn from_os_str(s: &OsStr) -> &Path {
+        unsafe { &*(s as *const OsStr as *const Path) }
```

Comment:
`osstr.as_ref() -> Path` already exists. Imo it's better to wait until that can be made const before introducing additional API surface for the sole purpose of supporting constification.